invoice/src/main/java/org/killbill/billing/invoice/InvoiceListener.java

```diff
@@ -31,114 +31,187 @@
+import org.killbill.billing.platform.api.LifecycleHandlerType.LifecycleLevel;
 import org.killbill.billing.subscription.api.SubscriptionBaseTransitionType;
 import org.killbill.billing.util.callcontext.CallOrigin;
 import org.killbill.billing.util.callcontext.InternalCallContextFactory;
 import org.killbill.billing.util.callcontext.UserType;
-import org.killbill.billing.util.config.definition.InvoiceConfig;
 import org.killbill.clock.Clock;
+import org.killbill.notificationq.api.NotificationQueueService;
+import org.killbill.notificationq.api.NotificationQueueService.NoSuchNotificationQueue;
+import org.killbill.queue.retry.RetryableService;
+import org.killbill.queue.retry.RetryableSubscriber;
+import org.killbill.queue.retry.RetryableSubscriber.SubscriberAction;
+import org.killbill.queue.retry.RetryableSubscriber.SubscriberQueueHandler;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.google.common.eventbus.AllowConcurrentEvents;
 import com.google.common.eventbus.Subscribe;
 import com.google.inject.Inject;
 
-public class InvoiceListener {
+@SuppressWarnings("TypeMayBeWeakened")
+public class InvoiceListener extends RetryableService implements InvoiceListenerService {
+
+    public static final String INVOICE_LISTENER_SERVICE_NAME = "invoice-listener-service";
 
     private static final Logger log = LoggerFactory.getLogger(InvoiceListener.class);
 
     private final InvoiceDispatcher dispatcher;
     private final InternalCallContextFactory internalCallContextFactory;
-    private final AccountInternalApi accountApi;
     private final InvoiceInternalApi invoiceApi;
-    private final InvoiceConfig invoiceConfig;
-    private final Clock clock;
+    private final RetryableSubscriber retryableSubscriber;
+    private final SubscriberQueueHandler subscriberQueueHandler = new SubscriberQueueHandler();
 
     @Inject
-    public InvoiceListener(final AccountInternalApi accountApi, final Clock clock, final InternalCallContextFactory internalCallContextFactory,
-                           final InvoiceConfig invoiceConfig, final InvoiceDispatcher dispatcher, InvoiceInternalApi invoiceApi) {
-        this.accountApi = accountApi;
+    public InvoiceListener(final AccountInternalApi accountApi,
+                           final InternalCallContextFactory internalCallContextFactory,
+                           final InvoiceDispatcher dispatcher,
+                           final InvoiceInternalApi invoiceApi,
+                           final NotificationQueueService notificationQueueService,
+                           final Clock clock) {
+        super(notificationQueueService);
         this.dispatcher = dispatcher;
-        this.invoiceConfig = invoiceConfig;
         this.internalCallContextFactory = internalCallContextFactory;
-        this.clock = clock;
         this.invoiceApi = invoiceApi;
+
+        subscriberQueueHandler.subscribe(EffectiveSubscriptionInternalEvent.class,
```

Comment:
Do we really need the retry layer from Kill Bill internal events ? Could we just plug that for plugins (the piece in platform `KillbillEventRetriableBusHandler`) ?

Reply:
These are separate pieces:

* `KillbillEventRetriableBusHandler` only takes care of retries for `ExtBusEvent` (notification plugins)
* The subscriber here is for invoice plugins (e.g. tax plugins) retries (original piece I worked on): the handler calls `processSubscriptionForInvoiceGeneration`, which eventually calls `invoicePluginDispatcher.getAdditionalInvoiceItems` (in `processAccountWithLockAndInputTargetDate`): this is the piece that can throw an `InvoicePluginApiRetryException`

Does that answer your question?